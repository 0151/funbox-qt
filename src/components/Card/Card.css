$SIN45: 0.68;

$width: 320px;
$height: 480px;

$border-width: 4px;
$border-radius: 12px;
$corner: 44px;

$text-secondary: #666;
$text-contrast: #fff;
$text-pseudolink-hover: #22a7e9;
$text-disabled: #ffff66;

$background: #f2f2f2;
$accent-default: #1698d9;
$accent-default-hover: #2ea8e6;
$accent-selected: #d91667;
$accent-selected-hover: #e52e7a;
$accent-disabled: #b3b3b3;

@define-mixin background-with-corner $color, $size: $corner {
  background-color: $color;
  background: linear-gradient(
    135deg,
    transparent 0,
    transparent calc($SIN45 * $size),
    $color calc($SIN45 * $size),
    $color 100%
  );
}

.card {
  display: inline-block;
  font-family: 'Trebuchet MS';
  width: $width;

  &__outer {
    box-sizing: border-box;
    padding: $border-width;
    @mixin background-with-corner $accent-default;
    cursor: pointer;
    border-radius: $border-radius;

    &:focus {
      outline: 0;
    }
  }

  &__inner {
    position: relative;
    z-index: 0;
    overflow: hidden;
    box-sizing: border-box;
    width: 100%;
    min-height: calc($height - 2 * $border-width);
    padding: 17px 48px;
    @mixin background-with-corner $background,
      calc($corner - 0.5 * $border-width);
    border-radius: calc($border-radius - $border-width);
  }

  &__subtitle {
    display: block;
    margin: 0 0 10px;
    color: $text-secondary;
    font-size: 16px;
    line-height: 19px;

    &_preventing {
      color: $accent-selected-hover;
    }
  }

  &__title {
    margin: 0;
  }

  &__name {
    display: block;
    margin-left: -3px;
    font-size: 48px;
    line-height: 46px;
    font-weight: bold;
  }

  &__taste {
    display: block;
    font-size: 24px;
    line-height: 36px;
    font-weight: bold;
  }

  &__features {
    color: $text-secondary;
    margin: 11px 0 0;
    padding: 0;
  }

  &__feature {
    font-size: 14px;
    line-height: 16px;
    list-style: none;
  }

  &__image {
    position: absolute;
    z-index: -1;
    display: block;
    height: 100%;
    width: 100%;
    bottom: 0;
    left: 0;
    background-image: url(./assets/cat.png);
    background-repeat: no-repeat;
    background-size: 368px 360px;
    background-position: 50% 182%;

    @media only screen and (min-device-pixel-ratio: 1.5) {
      background-image: url(./assets/cat@2x.png);
    }
  }

  &__footnote {
    display: block;
    margin-top: 14px;
    font-size: 13px;
    line-height: 15px;
    text-align: center;
    color: $text-contrast;
  }

  &__pseudolink {
    color: $accent-default;
    border-bottom: 1px dashed $accent-default;
    cursor: pointer;
  }

  &__dot {
    color: $accent-default;
  }

  &__weight {
    position: absolute;
    bottom: 12px;
    right: 12px;
    display: flex;
    box-sizing: border-box;
    height: 80px;
    width: 80px;
    padding: 2px 0 0 2px;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    color: $text-contrast;
    background-color: $accent-default;
    border-radius: 50%;
    line-height: 1;
    font-size: 42px;
  }

  &__unit {
    font-size: 21px;
  }

  &:hover {
    ^&__outer {
      @mixin background-with-corner $accent-default-hover;
    }

    ^&__weight {
      background-color: $accent-default-hover;
    }

    ^&__footnote {
      text-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
    }

    ^&__pseudolink {
      color: $text-pseudolink-hover;
    }

    ^&__dot {
      color: $accent-default-hover;
    }
  }

  /* Модификаторы */

  &_selected {
    ^&__outer {
      @mixin background-with-corner $accent-selected;
    }

    ^&__weight {
      background-color: $accent-selected;
    }
  }

  &_selected:hover {
    ^&__outer {
      @mixin background-with-corner $accent-selected-hover;
    }

    ^&__weight {
      background-color: $accent-selected-hover;
    }
  }

  &_disabled {
    pointer-events: none;

    ^&__outer {
      @mixin background-with-corner $accent-disabled;
    }

    ^&__subtitle,
    ^&__title,
    ^&__features {
      color: $accent-disabled;
      opacity: 0.5;
    }

    ^&__image {
      opacity: 0.5;
    }

    ^&__weight {
      background-color: $accent-disabled;
    }

    ^&__footnote {
      color: $text-disabled;
    }
  }
}
