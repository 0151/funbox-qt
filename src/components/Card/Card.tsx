import React, { FC, HTMLProps, useState } from 'react'

import { cn } from 'helpers/classname'

import './Card.css'

export interface ICardProps {
  /**
   * Вкус
   */
  taste: string
  /**
   * Особенности
   */
  features?: string[]
  /**
   * Описание
   */
  description: string
  /**
   * Вес
   */
  weight: number
  /**
   * Неактивное состояние
   */
  disabled?: boolean
}

const TEXT = {
  SUBTITLE: 'Сказочное заморское яство',
  PRODUCT_NAME: 'Нямушка',
  CALL_TO_ACTION: 'Чего сидишь? Порадуй котэ, ',
  BUY: 'купи',
  OUT_OF_STOCK: 'Печалька, %s закончился.',
  PREVENT_REMOVE: 'Котэ не одобряет?',
  KG: 'кг',
}

const b_ = cn('card')

export const Card: FC<ICardProps & HTMLProps<HTMLDivElement>> = ({
  taste,
  description,
  features,
  weight,
  disabled,
  className,
  ...restCardProps
}) => {
  const [selected, setSelected] = useState(false)
  const [hover, setHover] = useState(false)
  const preventing = hover && selected

  const handleClick = () => {
    setSelected(!selected)

    return selected || setHover(false)
  }

  const handleMouseHover = {
    onMouseEnter: () => selected && setHover(true),
    onMouseLeave: () => selected && setHover(false),
  }

  const handleKeyPress = (event: React.KeyboardEvent) => {
    event.preventDefault()

    if (event.key === ' ') {
      handleClick()
    }
  }

  return (
    <div className={b_({ selected, disabled }, [className])} {...restCardProps}>
      <div
        className={b_('outer')}
        role="checkbox"
        aria-checked={selected}
        tabIndex={disabled ? -1 : 1}
        onClick={handleClick}
        onKeyPress={handleKeyPress}
        {...handleMouseHover}
      >
        <div className={b_('inner')}>
          <span className={b_('subtitle', { preventing })}>
            {preventing ? TEXT.PREVENT_REMOVE : TEXT.SUBTITLE}
          </span>
          <h2 className={b_('title')}>
            <span className={b_('name')}>{TEXT.PRODUCT_NAME}</span>
            <span className={b_('taste')}>{taste}</span>
          </h2>
          <ul className={b_('features')}>
            {features?.map((feature) => (
              <li
                key={feature}
                className={b_('feature')}
                dangerouslySetInnerHTML={{ __html: feature }}
              />
            ))}
          </ul>
          <span className={b_('weight')}>
            {weight.toString().replace('.', ',')}
            <span className={b_('unit')}>{TEXT.KG}</span>
          </span>

          <span className={b_('image')} role="img" />
        </div>
      </div>
      <span className={b_('footnote')}>
        {selected && description}
        {disabled && TEXT.OUT_OF_STOCK.replace('%s', taste)}
        {!selected && !disabled && (
          <>
            {TEXT.CALL_TO_ACTION}
            <span
              className={b_('pseudolink')}
              role="button"
              tabIndex={0}
              onClick={handleClick}
              onKeyPress={handleKeyPress}
            >
              {TEXT.BUY}
            </span>
            <span className={b_('dot')}>.</span>
          </>
        )}
      </span>
    </div>
  )
}
