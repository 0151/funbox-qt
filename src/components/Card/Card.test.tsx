import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import _cards from 'data/cards.json'

import { Card } from './Card'
import type { ICardProps } from './Card'

const card: ICardProps = _cards[0]

describe('Карточка', () => {
  it('рендерится правильно', () => {
    const wrapper = shallow(<Card {...card} />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })
})
