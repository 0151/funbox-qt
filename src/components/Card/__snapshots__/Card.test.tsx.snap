// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Карточка рендерится правильно 1`] = `
<div
  className="card"
>
  <div
    aria-checked={false}
    className="card__outer"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="checkbox"
    tabIndex={1}
  >
    <div
      className="card__inner"
    >
      <span
        className="card__subtitle"
      >
        Сказочное заморское яство
      </span>
      <h2
        className="card__title"
      >
        <span
          className="card__name"
        >
          Нямушка
        </span>
        <span
          className="card__taste"
        >
          с фуа-гра
        </span>
      </h2>
      <ul
        className="card__features"
      >
        <li
          className="card__feature"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<b>10</b> порций",
            }
          }
          key="<b>10</b> порций"
        />
        <li
          className="card__feature"
          dangerouslySetInnerHTML={
            Object {
              "__html": "мышь в подарок",
            }
          }
          key="мышь в подарок"
        />
      </ul>
      <span
        className="card__weight"
      >
        0,5
        <span
          className="card__unit"
        >
          кг
        </span>
      </span>
      <span
        className="card__image"
        role="img"
      />
    </div>
  </div>
  <span
    className="card__footnote"
  >
    Чего сидишь? Порадуй котэ, 
    <span
      className="card__pseudolink"
      onClick={[Function]}
      role="button"
      tabIndex={0}
    >
      купи
    </span>
    <span
      className="card__dot"
    >
      .
    </span>
  </span>
</div>
`;
